functions/src/modules/CR.functions.ts

```diff
@@ -0,0 +1,96 @@
+import * as functions from 'firebase-functions';
+
+import * as helpers from '../helpers';
+import GoogleSheets from '../services/GoogleSheets';
+
+export const copyCRDataToProcessing = functions
+  .pubsub
+  .topic('daily-tick')
+  .onPublish(async (message) => {
+    const sourceSpreadsheet = new GoogleSheets(functions.config().cr.submissions.spreadsheet.id);
+    const onlineSheet = await sourceSpreadsheet.useSheet('Online');
+    const onlineSheetRows = await onlineSheet.getRows();
+
+    const destSpreadsheet = new GoogleSheets(functions.config().cr.processing.spreadsheet.id);
+    const destSpreadsheets = {};
+
+    const sheetNames = await destSpreadsheet.getSheetNames();
+
+    // used to hold `Submission subserials` of each sheet at the dest spreadsheet
+    const submissionSerials = {};
+
+    // Used to hold `Audio file names` of each sheet, to help look for the right place
+    // to insert a new submission
+    const audioFileNames = {};
+
+    // Holds the rows read from the `Online` sheet
+    // grouped by the sheet (list) name
+    const valuesToAppend = {};
+
+    for (const row of onlineSheetRows) {
+      const list = helpers.extractListFromFilename(row['Audio File Name']);
+
+      if (!destSpreadsheets[list]) {
+
+        // If the current sheet (list) doesn't exist on the target spreadsheet
+        if (sheetNames.indexOf(list) < 0) {
+          // 1. Create a new sheet with the curent list name
+          await destSpreadsheet.createSheet(list);
+
+          // 2. Create a new INSTANCE for the current sheet (list name)
+          destSpreadsheets[list] = await destSpreadsheet.useSheet(list);
+          // 3. Add the headers of the original sheet (ONLINE sheet)
+          await destSpreadsheets[list].appendRow(onlineSheet.headers, true);
+          sheetNames.push(list);
+        }
+      }
+
+      if (!submissionSerials[list]) {
+        submissionSerials[list] = await destSpreadsheets[list].getColumn('Submission Serial');
+        audioFileNames[list] = await destSpreadsheets[list].getColumn('Audio File Name');
+      }
+
+      if (!valuesToAppend[list])
+        valuesToAppend[list] = [];
+
+      valuesToAppend[list].push(row);
+    }
+
+
+    for (const sheetName in valuesToAppend) {
+      const sheetRows = valuesToAppend[sheetName];
+
+      const rows = [];
+      for (const row of sheetRows) {
+        const audioFileName = row['Audio File Name'];
+        const submissionSerial = row['Submission Serial'];
+
+
+        if (submissionSerials[sheetName].indexOf(submissionSerial) > -1)
+          continue;
+
+        let index = -1;
+        const sortedFileNames = audioFileNames[sheetName].sort();
+        for (let i = 0; i < sortedFileNames.length; i++) {
+          if (audioFileName < sortedFileNames[i]) {
+            index = i;
+            audioFileNames[sheetName] = [
```

Comment:
I understand the reason why you are trying to modify the `audioFileNames` array here, to adjust the index according to your inserted rows. But let's do it in another way.

- Keep original file names array intact.
- Get the original index for insertion by https://lodash.com/docs/4.17.11#sortedIndex.
- Of course, there may be other rows that have already been processed by you and targeted to be inserted before this index.
- In order to keep track of how many rows you are inserting at particular original index use the following construct: `insertions[index]++`.
- To find out the adjusted index for the current row, just sum up all the values in `insertions` array having index less or equal to you original index.

Should work out?